packages/flutter_tools/lib/src/test/flutter_web_platform.dart

```diff
@@ -702,3 +775,182 @@ class _BrowserEnvironment implements Environment {
   @override
   CancelableOperation<dynamic> displayPause() => _manager._displayPause();
 }
+
+/// Helper class to start golden file comparison in a separate process.
+///
+/// Golden file comparator is configured using flutter_test_config.dart and that
+/// file can contain arbitrary Dart code that depends on dart:ui. Thus it has to
+/// be executed in a `flutter_tester` environment. This helper class generates a
+/// Dart file configured with flutter_test_config.dart to perform the comparison
+/// of golden files.
+class TestGoldenComparator {
+  /// Creates a [TestGoldenComparator] instance.
+  TestGoldenComparator(this.shellPath, this.compilerFactory)
+      : tempDir = fs.systemTempDirectory.createTempSync('flutter_web_platform.');
+
+  final String shellPath;
+  final Directory tempDir;
+  final TestCompiler Function() compilerFactory;
+
+  TestCompiler _compiler;
+  TestGoldenComparatorProcess _previousComparator;
+  Uri _previousTestUri;
+
+  Future<void> close() async {
+    tempDir.deleteSync(recursive: true);
+    await _compiler?.dispose();
+    await _previousComparator?.close();
+  }
+
+  /// Start golden comparator in a separate process. Start one file per test file
+  /// to reduce the overhead of starting `flutter_tester`.
+  Future<TestGoldenComparatorProcess> _processForTestFile(Uri testUri) async {
+    if (testUri == _previousTestUri) {
+      return _previousComparator;
+    }
+
+    final String bootstrap = TestGoldenComparatorProcess.generateBootstrap(testUri);
+    final Process process = await _startProcess(bootstrap);
+    unawaited(_previousComparator?.close());
```

Comment:
Does this work in the presence of test concurrency? Or does that not work for golden files?

Reply:
I noticed that we have `--concurrency=1` for web tests on cirrus so I didn't take that into account when writing this.

Do we actually run tests concurrently for web? If so I can try to fix this.

Reply:
Just talked to @yjbanov , the test runner actually uses the same browser to run multiple tests in multiple iframes, so screenshot taking doesn't work in this case.